src/system_compositor.cpp

```diff
@@ -99,7 +123,24 @@ void usc::SystemCompositor::run()
                     mir::AbnormalExit("Video driver is blacklisted, exiting"));
             }
 
-            dm_connection = server->the_dm_connection();
+            if (server->get_options()->is_set("from-dm-fd") &&
+                server->get_options()->is_set("to-dm-fd"))
+            {
+                dm_connection = std::make_shared<usc::AsioDMConnection>(
+                    server->get_options()->get("from-dm-fd", -1),
+                    server->get_options()->get("to-dm-fd", -1),
+                    session_switcher);
+            }
+            else if (server->get_options()->is_set("debug-without-dm"))
+            {
+                dm_connection = std::make_shared<NullDMMessageHandler>(
```

Comment:
This patch introduces a segfault when running `unity-system-compositor --debug-without-dm`.  Without having looked in depth yet, my guess would be that `session_switcher` is not yet initialized, probably because the init callback is called before the callback provided to `override_the_window_manager_builder`.

Probably the best way to handle this is to provide `usc::Server` and `usc::SystemCompositor` with a callback that lazily initializes `session_switcher`, instead of just providing the `shared_ptr` directly.